File: 0155-min-stack/0155-min-stack.js
var Node = function(val, next) {
    this.val = val;
    this.next = next;
}

var MinStack = function() {
    this.minVal = Number.MAX_SAFE_INTEGER;
    this.topmost = null;
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    let valToInsert = val;
    if (val < this.minVal) {
        if (this.topmost !== null) {
            valToInsert = 2 * val - this.minVal;
        }
        this.minVal = val;
    }
    const node = new Node(valToInsert, this.topmost);
    this.topmost = node;
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    if (this.topmost.val < this.minVal) {
        this.minVal = 2 * this.minVal - this.topmost.val;
    }
    this.topmost = this.topmost.next;
    if (this.topmost === null) {
        this.minVal = Number.MAX_SAFE_INTEGER;
    }
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.topmost.val < this.minVal ? this.minVal : this.topmost.val;
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
   return this.minVal; 
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: 0155-min-stack/0155-min-stack.cpp
class StackNode {
public:
    long val;
    StackNode *next;
    
    StackNode(long x, StackNode *node) {
        val = x;
        next = node;
    }
};

class MinStack {
    StackNode *topMost;
    int minVal;
public:
    MinStack() {
        topMost = NULL;
        minVal = INT_MAX;
    }
    
    void push(int val) {
        long valToInsert = val;
        if (val < minVal) {
            if (topMost != NULL) valToInsert = 2 * (long)val - minVal;
            minVal = val;
        }
        StackNode *node = new StackNode(valToInsert, topMost);
        topMost = node;
    }
    
    void pop() {
        if (topMost->val < minVal) minVal = 2 * (long)minVal - topMost->val;
        StackNode *temp = topMost;
        topMost = topMost->next;
        delete temp;

        if (topMost == NULL) minVal = INT_MAX;
    }
    
    int top() {
        return topMost->val < minVal ? minVal : topMost->val;
    }
    
    int getMin() {
        return minVal;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */